skills/repo.go

```diff
@@ -19,10 +19,200 @@
+type Repository struct {
+	// Protects the internals from concurrent access.
+	m      sync.Mutex
+	types  []Type
+	skills []*Skill
+}
+
+var (
+	// ErrTypeNotFound is reported when skill type cannot found.
+	ErrTypeNotFound = errors.New("skill type not found")
+	// ErrDuplicateType is reported when type with duplicate name is being added to a repository.
+	ErrDuplicateType = errors.New("duplicate type name")
+	// ErrSkillNotFound is reported when skill cannot be looked up.
+	ErrSkillNotFound = errors.New("skill not found")
+	// ErrDuplicateSkill is reported when skill with duplicate name is being added to a repository.
+	ErrDuplicateSkill = errors.New("duplicate skill name")
+)
 
 // NewRepository creates an empty skill repository.
 func NewRepository() *Repository {
 	return &Repository{}
 }
+
+// AllTypes returns all skill types known to the repository.
+func (r *Repository) AllTypes() []Type {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	return append([]Type(nil), r.types...)
+}
+
+// Type returns a type with a given name.
+func (r *Repository) Type(typeName string) Type {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	return r.unlockedType(typeName)
+}
+
+// AddType adds the provided skill type to the repository.
+func (r *Repository) AddType(t Type) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	typeName := t.Name()
+	if err := ValidateName(typeName); err != nil {
+		return err
+	}
+	if i, found := r.unlockedTypeIndex(typeName); !found {
+		r.types = append(r.types[:i], append([]Type{t}, r.types[i:]...)...)
+		return nil
+	}
+	return ErrDuplicateType
+}
+
+// AllSkills returns all skills of the given type.
+// If skillType is the empty string, all skills are returned.
+func (r *Repository) AllSkills(skillType string) []*Skill {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	var result []*Skill
+	if skillType == "" {
+		result = make([]*Skill, len(r.skills))
+		copy(result, r.skills)
+	} else {
+		result = make([]*Skill, 0)
+		for _, skill := range r.skills {
+			if skill.Type == skillType {
+				result = append(result, skill)
+			}
+		}
+	}
+	return result
+}
+
+// Skills returns the skills offered by the named snap.
+func (r *Repository) Skills(snapName string) []*Skill {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	var result []*Skill
+	for _, skill := range r.skills {
+		if skill.Snap == snapName {
+			result = append(result, skill)
+		}
+	}
+	return result
+}
+
+// Skill returns the specified skill from the named snap.
+func (r *Repository) Skill(snapName, skillName string) *Skill {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	return r.unlockedSkill(snapName, skillName)
+}
+
+// AddSkill adds a skill to the repository.
+// Skill names must be valid snap names, as defined by ValidateName.
+// Skill name must be unique within a particular snap.
+func (r *Repository) AddSkill(snapName, skillName, typeName, label string, attrs map[string]interface{}) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	// Reject skill with invalid names
+	if err := ValidateName(snapName); err != nil {
```

Comment:
That's a good point. I'm not sure we do (grepping quickly I didn't find one). Since we don't really look at snap names much, perhaps we don't really need to validate it?